### PR TITLE
Fix non contiguous co reduction on 2+D arrays

### PR DESCRIPTION
**Pull request by neok-m4700**
[links]:#
[contributing guidelines]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/CONTRIBUTING.md
[issue]: https://github.com/sourceryinstitute/OpenCoarrays/issues
[PR response img]: https://img.shields.io/issuestats/p/github/sourceryinstitute/OpenCoarrays.svg?style=flat-square
[coverage]: https://img.shields.io/codecov/c/github/sourceryinstitute/OpenCoarrays/master.svg?style=flat-square

|  Avg response time                |  coverage on master         |
|:---------------------------------:|:---------------------------:|
| ![Issue Stats][PR response img]   | ![Codecov branch][coverage] |

## Summary of changes ##

This is a fix for co reduction on non contiguous  ndarrays (dim > 2).
Logic of extents has been simplified.
A specific regession test related to the work done in #503 is added.
Removed 2 broken symlinks (src/tests/regression/open).

@zbeekman : for the next issues, should I put new tests under src/tests/regression/reported, even if I don't open a specific issue and instead PR a new patch + test case ? Maybe create a new directory under src/tests/regression ?

## Additional info and certifications ##

This pull request (PR) is a:

- [x] Bug fix
- [ ] Feature addition
- [ ] Other, Please describe:

### I certify that ###

- [x] I reviewed and followed the [contributing guidelines], including
      - Increasing test coverage for all feature-addition PRs
      - Increasing test coverage for all bug-fix PRs for which there
        does not already exist a related test that failed before the PR
      - At least maintaining test coverage for all other PRs
      - Ensuring that all tests pass when run locally 
      - Naming PR  to indicate work in progress (WIP) and to attach the PR
        to the appropriate bug report or feature request [issue]
      - White space (no trailing white space or white space errors may
        be introduced)
      - Commenting code where it is non-obvious and non-trivial
      - Logically atomic, self consistent and coherent commits
      - Commit message content
      - Waiting 24 hours before self-approving the PR to give another
        OpenCoarrays developer a chance to review my proposed code

### CODE REVIEW

**Comment by rouson** [comment]
@neok-m4700 Thanks for your contributions.  Feel free to update the PR with the new test directory you're suggesting.  Zaak is on vacation this week so I'll be the one to merge this.  All tests pass so I hope to look at the commits and approve and merge all 3 of your recently submitted PRs within the next 48-72 hours.

**Comment by neok-m4700** [comment]
Thanks for the update @rouson,
Better leave some time so that one can review these PRs before merging. Quality first !

**Comment by rouson** [comment]
LGTM

**Comment by zbeekman** [comment]
> Removed 2 broken symlinks (src/tests/regression/open).

I only see changes for 1 symlink... Am I missing something?

> @zbeekman : for the next issues, should I put new tests under src/tests/regression/reported, even if I don't open a specific issue and instead PR a new patch + test case ? Maybe create a new directory under src/tests/regression ?

I'm open to either idea, but my preference would be to create a brief new issue, and then put the test in `src/rests/regression/reported`. If you have code submitted you can skimp on the description of the issue a bit, and I won't be as pedantic about the issue form---i.e., skip all the platform/compiler/environment data. The submitted test will be much more concrete documentation than info in the issue.

File: src/mpi/mpi_caf.c [hunk 1]
```diff
@@ -6787,7 +6787,7 @@ internal_co_reduce (MPI_Op op, gfc_descriptor_t *source, int result_image, int *
   MPI_Datatype datatype = get_MPI_datatype (source, src_len);
 
   size = 1;
-  for (j = 0; j < rank; j++)
+  for (j = 0; j < rank; ++j)
```
[the remark this answers is not in the document]
Comment (neok-m4700) [reply src/mpi/mpi_caf.c]: :wink: this guy escaped !

**Comment by neok-m4700** [comment]
@zbeekman 
Ok, I will create an issue first, and propose a PR afterwards, I was afraid this might be redundant, but it's definitively better for VC. Will do (and truncate the headers).

**Comment by zbeekman** [comment]
If you contribute a regression test (which we are always very grateful for 🙏) with the PR then please just open a very brief issue. You can nuke all/most of the template and just describe the bug/regression. This helps track things.

If you open a more minor PR (formatting, documentation, etc.) and no test is contributed, then no need to open an issue, if it is truly very minor.

**Pull request merged**